pandas/core/reshape.py

```diff
@@ -1153,6 +1175,48 @@ def get_dummies(data, prefix=None, prefix_sep='_', dummy_na=False,
     from pandas.tools.concat import concat
     from itertools import cycle
 
+    # Deprecate filling NaN values with zeros, GH15926
+    # When this is finally deprecated, simply remove this block
+    # of code and change the default to np.nan in the function signature
+    # of `get_dummies`.
+    if fill_value is None:
+        warnings.warn('The default behavior of filling NaN values '
+                      'with zeros will be deprecated. Use '
+                      '`df = pd.get_dummies(df, fill_value=0)` to reproduce '
+                      'this behavior', FutureWarning, 3)
+        fill_value = 0
+
+    # Infer the proper output dtype.
+    # GH15926
+    try:
+        if np.all(np.isfinite(data.values if hasattr(data, 'values') else data)):
+            any_null = True
+        else:
+            any_null = False
+    except TypeError:
+        any_null = False
+
+    if any_null:
+        output_dtype = np.uint8
+    else:
+        fill_value_dtype, fill_value = infer_dtype_from_scalar(fill_value)
+
+        if 'int' in str(fill_value_dtype):
+            if fill_value >= 0:
+                if fill_value <= np.iinfo(np.uint8).max:
+                    output_dtype = np.uint8
+                else:
+                    output_dtype = np.uint64
+            else:
+                if fill_value >= np.iinfo(np.int8).min and fill_value <= np.iinfo(np.int8).max:
+                    output_dtype = np.int8
+                else:
+                    output_dtype = np.int64
+        elif 'float' in str(fill_value_dtype):
+            output_dtype = np.float32
+        else:
+            raise ValueError('`fill_value` must be `np.nan`, an int or a float type!')
```

Comment:
yep all for that! (something like this you can do here / or in another PR).

an alternative is to strip out the dtype inference and do it in another PR (before this one). That makes this much cleaner.